### ISSUE TITLE
[RFC] SemVer contract for API in 2.0?

### ISSUE THREAD

**Issue by davidmehren** [codimd/server#399]
In #396 we stumbled upon the question if we should support 3rd party clients and provide a stable API for them. I would like to gather a bit more feedback and thoughts on that topic.

Currently, we don't have a documented API. This is going to change with 2.0 and we could publish the docs for that with the expectation that third party clients will use it. Of course that would add some burden to us, as we need to take care to always do a corresponding SemVer update when the API is changed. That would then allow other software to rely on our API and integrate it.

We could also publish the API docs with a "no stability guarantees" warning, which would free us from that burden. This would mean that only our own official frontends need to be compatible with API changes and we would just need to make sure we release a new version of all of them alongside a new backend. Our SemVer would then only apply to user-facing changes.

Personally, I would love an ecosystem of third party clients or integrations and think we should think about making our API 'really public'.

**Comment by SISheogorath** [comment]
My idea about the API would be to provide a super stable API across multiple CodiMD versions for the most commonly requested actions (creating a note, changing permissions, …) for third party clients and probably having a less stable API for tight integration with the frontend, that can be adjusted to the frontends needs.

The frontend in 1.X was thightly integrated with the backend including SSR, which means the API that was envisioned for 2.0 was mainly for exactly the most common actions, such as updating notes, creating notes, receiving history, etc.

While the current react-client tries to build a full-featured client against this API, which I personally think, will make this API too unstable for traditional third-party integrations or requires a lot of engineering which will start to limit us down the road.

It might be an idea, that we create a 0.2.x API endpoint that will go forward as backend and frontend develop and will not provide and backwards compatibility and provide an LTS API like 2.0 which provides the features mentioned in the beginning.

The reasons are quite simple: I'm 100% sure, that for the next year, the frontend we deliver will be the only full featured frontend for CodiMD. While a lot of scripts will appear surrounding the API and will break every time we want to rename a field or add something new.

For the 2.0 API due to the reduced set of capabilities the contract can be much stricter without limiting client implementations.

**Comment by DerMolly** [comment]
So am I right that you suggesting to have two APIs with more or less the same functionality? One which we won't change as much (you called it LTS API) and another one which can be more easily changed in upcoming releases?
Or do you mean to separate the API in a stable LTS part and and a more rapidly moving non-LTS part. If that's the idea, where do we make the cut. What goes where?

**Comment by SISheogorath** [comment]
Yes, two APIs that provide more or less the same functionality. The LTS API should be a bit lighter from a feature perspective (i.e. no realtime features, no configs, nothing frontend specific, …) It should focus on use-cases that one might would expect from a CLI using curl or from something people integrate in their third-party business application that gets released once or twice a year.

And the unstable API is an internal piece of work, that we can break as we need it and this way ensure less headache about further development for everyone involved.

**Comment by SISheogorath** [comment]
@codimd/frontend agreed that it would be useful to define a LTS API with a limited set of features which we started to brainstorm here:

- GET /notes (all notes)
- POST /notes (create new note)
- GET /notes/{noteId}
- PUT /notes/{noteId} (does the follow actions: creates revision, updates note, disconnects all clients)
- GET /notes/{noteId}/permissions
- PUT /notes/{noteId}/permissions
- GET /me/export (to be renamed)
- GET /notes/{noteId}/revisions
- GET /notes/{noteId}/revisions/{revisionId}
- GET /notes/{noteId}/revisions/{revisionId}/content
- GET /notes/{noteId}/content
- GET /notes/{noteId}/info
- GET /monitoring (previous `/status`) (possibly multiple endpoint under /monitoring)

It should be specified in more detailed in a proper API documentation. Implementation details such a JSON Schemas and alike have to be done later on.

If you think there are additional endpoint you need or if you have questions about what endpoint is supposed to do what, because I don't get it from a name, feel free throw a comment :slightly_smiling_face:
